File: bench/random_crossval.zig
const bcs = @import("bcs");
const std = @import("std");

comptime {
    @setEvalBranchQuota(10_000);
}

const ser_cases = 128;
const de_cases = 256;
const seed: u64 = 0x5eed_cafe_d00d_f00d;
const ascii = "abcxyz012_";

const FuzzStruct = struct {
    a: u64,
    b: bool,
    c: []const u16,
    d: ?[4]u8,
};

const FuzzEnum = union(enum) {
    Unit,
    U64: u64,
    Bytes: [4]u8,
    Pair: struct { a: u16, b: u8 },
};

const DeepStruct = struct {
    label: bcs.String,
    nested: FuzzStruct,
    variant: FuzzEnum,
    tags: []const bcs.String,
};

const Rng = struct {
    state: u64,

    fn init(s: u64) Rng {
        return .{ .state = s };
    }

    fn nextU64(self: *Rng) u64 {
        var x = self.state;
        x ^= x >> 12;
        x ^= x << 25;
        x ^= x >> 27;
        self.state = x;
        return x *% 0x2545_F491_4F6C_DD1D;
    }

    fn nextBool(self: *Rng) bool {
        return self.nextU64() & 1 == 1;
    }

    fn nextU8(self: *Rng) u8 {
        return @truncate(self.nextU64());
    }

    fn nextU16(self: *Rng) u16 {
        return @truncate(self.nextU64());
    }

    fn nextU32(self: *Rng) u32 {
        return @truncate(self.nextU64());
    }

    fn range(self: *Rng, limit: usize) usize {
        if (limit == 0) return 0;
        return @intCast(self.nextU64() % limit);
    }
};

fn emitBytes(name: []const u8, bytes: []const u8) void {
    std.debug.print("{s}=", .{name});
    for (bytes) |byte| std.debug.print("{x:0>2}", .{byte});
    std.debug.print("\n", .{});
}

fn emitSerialized(name: []const u8, value: anytype) !void {
    const allocator = std.heap.c_allocator;
    const bytes = try bcs.serialize(allocator, value);
    defer allocator.free(bytes);
    emitBytes(name, bytes);
}

fn emitStatus(comptime T: type, name: []const u8, raw: []const u8) void {
    const allocator = std.heap.c_allocator;
    const value = bcs.deserialize(T, allocator, raw) catch {
        std.debug.print("{s}=err\n", .{name});
        return;
    };
    defer bcs.freeDeserialized(T, allocator, value);

    const encoded = bcs.serialize(allocator, value) catch {
        std.debug.print("{s}=err\n", .{name});
        return;
    };
    defer allocator.free(encoded);
    std.debug.print("{s}=ok:", .{name});
    for (encoded) |byte| std.debug.print("{x:0>2}", .{byte});
    std.debug.print("\n", .{});
}

fn allocBytes(allocator: std.mem.Allocator, rng: *Rng, max_len: usize) ![]const u8 {
    const len = rng.range(max_len + 1);
    const bytes = try allocator.alloc(u8, len);
    for (bytes) |*byte| byte.* = rng.nextU8();
    return bytes;
}

fn allocAsciiString(allocator: std.mem.Allocator, rng: *Rng, max_len: usize) !bcs.String {
    const len = rng.range(max_len + 1);
    const bytes = try allocator.alloc(u8, len);
    for (bytes) |*byte| byte.* = ascii[rng.range(ascii.len)];
    return bcs.String.init(bytes);
}

fn allocU16Vec(allocator: std.mem.Allocator, rng: *Rng, max_len: usize) ![]const u16 {
    const len = rng.range(max_len + 1);
    const values = try allocator.alloc(u16, len);
    for (values) |*value| value.* = rng.nextU16();
    return values;
}

fn allocVecVecU8(allocator: std.mem.Allocator, rng: *Rng, max_outer: usize, max_inner: usize) ![]const []const u8 {
    const len = rng.range(max_outer + 1);
    const values = try allocator.alloc([]const u8, len);
    for (values) |*value| value.* = try allocBytes(allocator, rng, max_inner);
    return values;
}

fn allocStringVec(allocator: std.mem.Allocator, rng: *Rng, max_len: usize, max_str_len: usize) ![]const bcs.String {
    const len = rng.range(max_len + 1);
    const values = try allocator.alloc(bcs.String, len);
    for (values) |*value| value.* = try allocAsciiString(allocator, rng, max_str_len);
    return values;
}

fn genOptU64(rng: *Rng) ?u64 {
    if (rng.nextBool()) return rng.nextU64();
    return null;
}

fn genArr4(rng: *Rng) [4]u8 {
    return .{ rng.nextU8(), rng.nextU8(), rng.nextU8(), rng.nextU8() };
}

fn allocFuzzStruct(allocator: std.mem.Allocator, rng: *Rng) !FuzzStruct {
    return .{
        .a = rng.nextU64(),
        .b = rng.nextBool(),
        .c = try allocU16Vec(allocator, rng, 6),
        .d = if (rng.nextBool()) genArr4(rng) else null,
    };
}

fn allocDeepStruct(allocator: std.mem.Allocator, rng: *Rng) !DeepStruct {
    return .{
        .label = try allocAsciiString(allocator, rng, 10),
        .nested = try allocFuzzStruct(allocator, rng),
        .variant = genFuzzEnum(rng),
        .tags = try allocStringVec(allocator, rng, 5, 6),
    };
}

fn genFuzzEnum(rng: *Rng) FuzzEnum {
    return switch (rng.range(4)) {
        0 => .Unit,
        1 => .{ .U64 = rng.nextU64() },
        2 => .{ .Bytes = genArr4(rng) },
        else => .{ .Pair = .{ .a = rng.nextU16(), .b = rng.nextU8() } },
    };
}

fn allocMapU8U16(allocator: std.mem.Allocator, rng: *Rng, max_len: usize) !bcs.Map(u8, u16) {
    const M = bcs.Map(u8, u16);
    const len = rng.range(max_len + 1);
    const entries = try allocator.alloc(M.Entry, len);
    for (entries) |*entry| {
        entry.* = .{
            .key = rng.nextU8() % 5,
            .value = rng.nextU16(),
        };
    }
    return M.from(entries);
}

fn allocMapStringU32(allocator: std.mem.Allocator, rng: *Rng, max_len: usize) !bcs.Map(bcs.String, u32) {
    const M = bcs.Map(bcs.String, u32);
    const len = rng.range(max_len + 1);
    const entries = try allocator.alloc(M.Entry, len);
    for (entries) |*entry| {
        entry.* = .{
            .key = try allocAsciiString(allocator, rng, 4),
            .value = rng.nextU32(),
        };
    }
    return M.from(entries);
}

pub fn main() !void {
    var ser_rng = Rng.init(seed);

    for (0..ser_cases) |i| {
        {
            var name_buf: [32]u8 = undefined;
            const name = try std.fmt.bufPrint(&name_buf, "ser_bool_{d:0>3}", .{i});
            try emitSerialized(name, ser_rng.nextBool());
        }
        {
            var name_buf: [32]u8 = undefined;
            const name = try std.fmt.bufPrint(&name_buf, "ser_u64_{d:0>3}", .{i});
            try emitSerialized(name, ser_rng.nextU64());
        }
        {
            var name_buf: [32]u8 = undefined;
            const name = try std.fmt.bufPrint(&name_buf, "ser_i64_{d:0>3}", .{i});
            try emitSerialized(name, @as(i64, @bitCast(ser_rng.nextU64())));
        }
        {
            var arena = std.heap.ArenaAllocator.init(std.heap.c_allocator);
            defer arena.deinit();
            var name_buf: [32]u8 = undefined;
            const name = try std.fmt.bufPrint(&name_buf, "ser_string_{d:0>3}", .{i});
            try emitSerialized(name, try allocAsciiString(arena.allocator(), &ser_rng, 12));
        }
        {
            var arena = std.heap.ArenaAllocator.init(std.heap.c_allocator);
            defer arena.deinit();
            var name_buf: [32]u8 = undefined;
            const name = try std.fmt.bufPrint(&name_buf, "ser_bytes_{d:0>3}", .{i});
            try emitSerialized(name, try allocBytes(arena.allocator(), &ser_rng, 16));
        }
        {
            var arena = std.heap.ArenaAllocator.init(std.heap.c_allocator);
            defer arena.deinit();
            var name_buf: [32]u8 = undefined;
            const name = try std.fmt.bufPrint(&name_buf, "ser_vec_u16_{d:0>3}", .{i});
            try emitSerialized(name, try allocU16Vec(arena.allocator(), &ser_rng, 8));
        }
        {
            var arena = std.heap.ArenaAllocator.init(std.heap.c_allocator);
            defer arena.deinit();
            var name_buf: [32]u8 = undefined;
            const name = try std.fmt.bufPrint(&name_buf, "ser_vec_vec_u8_{d:0>3}", .{i});
            try emitSerialized(name, try allocVecVecU8(arena.allocator(), &ser_rng, 4, 8));
        }
        {
            var name_buf: [32]u8 = undefined;
            const name = try std.fmt.bufPrint(&name_buf, "ser_opt_u64_{d:0>3}", .{i});
            try emitSerialized(name, genOptU64(&ser_rng));
        }
        {
            var arena = std.heap.ArenaAllocator.init(std.heap.c_allocator);
            defer arena.deinit();
            var name_buf: [32]u8 = undefined;
            const name = try std.fmt.bufPrint(&name_buf, "ser_struct_{d:0>3}", .{i});
            try emitSerialized(name, try allocFuzzStruct(arena.allocator(), &ser_rng));
        }
        {
            var name_buf: [32]u8 = undefined;
            const name = try std.fmt.bufPrint(&name_buf, "ser_enum_{d:0>3}", .{i});
            try emitSerialized(name, genFuzzEnum(&ser_rng));
        }
        {
            var arena = std.heap.ArenaAllocator.init(std.heap.c_allocator);
            defer arena.deinit();
            var name_buf: [36]u8 = undefined;
            const name = try std.fmt.bufPrint(&name_buf, "ser_map_u8_u16_{d:0>3}", .{i});
            try emitSerialized(name, try allocMapU8U16(arena.allocator(), &ser_rng, 6));
        }
        {
            var arena = std.heap.ArenaAllocator.init(std.heap.c_allocator);
            defer arena.deinit();
            var name_buf: [40]u8 = undefined;
            const name = try std.fmt.bufPrint(&name_buf, "ser_map_string_u32_{d:0>3}", .{i});
            try emitSerialized(name, try allocMapStringU32(arena.allocator(), &ser_rng, 6));
        }
        {
            var arena = std.heap.ArenaAllocator.init(std.heap.c_allocator);
            defer arena.deinit();
            var name_buf: [40]u8 = undefined;
            const name = try std.fmt.bufPrint(&name_buf, "ser_deep_struct_{d:0>3}", .{i});
            try emitSerialized(name, try allocDeepStruct(arena.allocator(), &ser_rng));
        }
    }

    var de_rng = Rng.init(seed ^ 0x9e37_79b9_7f4a_7c15);

    for (0..de_cases) |i| {
        var raw_buf: [24]u8 = undefined;
        const raw_len = de_rng.range(raw_buf.len + 1);
        for (raw_buf[0..raw_len]) |*byte| byte.* = de_rng.nextU8();
        const raw = raw_buf[0..raw_len];

        {
            var name_buf: [32]u8 = undefined;
            const name = try std.fmt.bufPrint(&name_buf, "de_bool_{d:0>3}", .{i});
            emitStatus(bool, name, raw);
        }
        {
            var name_buf: [32]u8 = undefined;
            const name = try std.fmt.bufPrint(&name_buf, "de_opt_u8_{d:0>3}", .{i});
            emitStatus(?u8, name, raw);
        }
        {
            var name_buf: [32]u8 = undefined;
            const name = try std.fmt.bufPrint(&name_buf, "de_string_{d:0>3}", .{i});
            emitStatus(bcs.String, name, raw);
        }
        {
            var name_buf: [32]u8 = undefined;
            const name = try std.fmt.bufPrint(&name_buf, "de_vec_u8_{d:0>3}", .{i});
            emitStatus([]const u8, name, raw);
        }
        {
            var name_buf: [32]u8 = undefined;
            const name = try std.fmt.bufPrint(&name_buf, "de_vec_u16_{d:0>3}", .{i});
            emitStatus([]const u16, name, raw);
        }
        {
            var name_buf: [32]u8 = undefined;
            const name = try std.fmt.bufPrint(&name_buf, "de_enum_{d:0>3}", .{i});
            emitStatus(FuzzEnum, name, raw);
        }
        {
            var name_buf: [36]u8 = undefined;
            const name = try std.fmt.bufPrint(&name_buf, "de_tuple_pair_{d:0>3}", .{i});
            emitStatus(struct { u32, bool }, name, raw);
        }
        {
            var name_buf: [40]u8 = undefined;
            const name = try std.fmt.bufPrint(&name_buf, "de_map_u8_unit_{d:0>3}", .{i});
            emitStatus(bcs.Map(u8, void), name, raw);
        }
        {
            var name_buf: [36]u8 = undefined;
            const name = try std.fmt.bufPrint(&name_buf, "de_struct_{d:0>3}", .{i});
            emitStatus(FuzzStruct, name, raw);
        }
        {
            var name_buf: [40]u8 = undefined;
            const name = try std.fmt.bufPrint(&name_buf, "de_deep_struct_{d:0>3}", .{i});
            emitStatus(DeepStruct, name, raw);
        }
        {
            var name_buf: [36]u8 = undefined;
            const name = try std.fmt.bufPrint(&name_buf, "de_map_u8_u16_{d:0>3}", .{i});
            emitStatus(bcs.Map(u8, u16), name, raw);
        }
        {
            var name_buf: [40]u8 = undefined;
            const name = try std.fmt.bufPrint(&name_buf, "de_map_string_u32_{d:0>3}", .{i});
            emitStatus(bcs.Map(bcs.String, u32), name, raw);
        }
    }
}

File: bench/random_crossval.rs
use serde::{ser::SerializeMap, Deserialize, Serialize};
use std::collections::BTreeMap;

const SER_CASES: usize = 128;
const DE_CASES: usize = 256;
const SEED: u64 = 0x5eed_cafe_d00d_f00d;
const ASCII: &[u8] = b"abcxyz012_";

#[derive(Clone, Debug, Serialize, Deserialize)]
struct FuzzStruct {
    a: u64,
    b: bool,
    c: Vec<u16>,
    d: Option<[u8; 4]>,
}

#[derive(Clone, Debug, Serialize, Deserialize)]
enum FuzzEnum {
    Unit,
    U64(u64),
    Bytes([u8; 4]),
    Pair { a: u16, b: u8 },
}

#[derive(Clone, Debug, Serialize, Deserialize)]
struct DeepStruct {
    label: String,
    nested: FuzzStruct,
    variant: FuzzEnum,
    tags: Vec<String>,
}

#[derive(Clone, Debug)]
struct EntryMapU8U16 {
    entries: Vec<(u8, u16)>,
}

impl Serialize for EntryMapU8U16 {
    fn serialize<S>(&self, serializer: S) -> Result<S::Ok, S::Error>
    where
        S: serde::Serializer,
    {
        let mut map = serializer.serialize_map(Some(self.entries.len()))?;
        for (k, v) in &self.entries {
            map.serialize_entry(k, v)?;
        }
        map.end()
    }
}

#[derive(Clone, Debug)]
struct EntryMapStringU32 {
    entries: Vec<(String, u32)>,
}

impl Serialize for EntryMapStringU32 {
    fn serialize<S>(&self, serializer: S) -> Result<S::Ok, S::Error>
    where
        S: serde::Serializer,
    {
        let mut map = serializer.serialize_map(Some(self.entries.len()))?;
        for (k, v) in &self.entries {
            map.serialize_entry(k, v)?;
        }
        map.end()
    }
}

#[derive(Clone, Debug)]
struct Rng {
    state: u64,
}

impl Rng {
    fn new(seed: u64) -> Self {
        Self { state: seed }
    }

    fn next_u64(&mut self) -> u64 {
        let mut x = self.state;
        x ^= x >> 12;
        x ^= x << 25;
        x ^= x >> 27;
        self.state = x;
        x.wrapping_mul(0x2545_F491_4F6C_DD1D)
    }

    fn next_bool(&mut self) -> bool {
        self.next_u64() & 1 == 1
    }

    fn next_u8(&mut self) -> u8 {
        self.next_u64() as u8
    }

    fn next_u16(&mut self) -> u16 {
        self.next_u64() as u16
    }

    fn next_u32(&mut self) -> u32 {
        self.next_u64() as u32
    }

    fn range(&mut self, limit: usize) -> usize {
        if limit == 0 {
            0
        } else {
            (self.next_u64() % limit as u64) as usize
        }
    }
}

fn hex(bytes: &[u8]) -> String {
    bytes
        .iter()
        .map(|b| format!("{:02x}", b))
        .collect::<Vec<_>>()
        .join("")
}

fn emit_bytes(name: &str, bytes: &[u8]) {
    println!("{name}={}", hex(bytes));
}

fn emit_status<T>(name: &str, bytes: &[u8])
where
    T: Serialize + for<'de> Deserialize<'de>,
{
    match bcs::from_bytes::<T>(bytes) {
        Ok(value) => {
            let encoded = bcs::to_bytes(&value).unwrap();
            println!("{name}=ok:{}", hex(&encoded));
        }
        Err(_) => println!("{name}=err"),
    }
}

fn gen_bytes(rng: &mut Rng, max_len: usize) -> Vec<u8> {
    let len = rng.range(max_len + 1);
    (0..len).map(|_| rng.next_u8()).collect()
}

fn gen_ascii_string(rng: &mut Rng, max_len: usize) -> String {
    let len = rng.range(max_len + 1);
    let bytes = (0..len)
        .map(|_| ASCII[rng.range(ASCII.len())])
        .collect::<Vec<_>>();
    String::from_utf8(bytes).unwrap()
}

fn gen_u16_vec(rng: &mut Rng, max_len: usize) -> Vec<u16> {
    let len = rng.range(max_len + 1);
    (0..len).map(|_| rng.next_u16()).collect()
}

fn gen_vec_vec_u8(rng: &mut Rng, max_outer: usize, max_inner: usize) -> Vec<Vec<u8>> {
    let len = rng.range(max_outer + 1);
    (0..len).map(|_| gen_bytes(rng, max_inner)).collect()
}

fn gen_string_vec(rng: &mut Rng, max_len: usize, max_str_len: usize) -> Vec<String> {
    let len = rng.range(max_len + 1);
    (0..len)
        .map(|_| gen_ascii_string(rng, max_str_len))
        .collect()
}

fn gen_opt_u64(rng: &mut Rng) -> Option<u64> {
    if rng.next_bool() {
        Some(rng.next_u64())
    } else {
        None
    }
}

fn gen_arr4(rng: &mut Rng) -> [u8; 4] {
    [rng.next_u8(), rng.next_u8(), rng.next_u8(), rng.next_u8()]
}

fn gen_fuzz_struct(rng: &mut Rng) -> FuzzStruct {
    FuzzStruct {
        a: rng.next_u64(),
        b: rng.next_bool(),
        c: gen_u16_vec(rng, 6),
        d: if rng.next_bool() {
            Some(gen_arr4(rng))
        } else {
            None
        },
    }
}

fn gen_deep_struct(rng: &mut Rng) -> DeepStruct {
    DeepStruct {
        label: gen_ascii_string(rng, 10),
        nested: gen_fuzz_struct(rng),
        variant: gen_fuzz_enum(rng),
        tags: gen_string_vec(rng, 5, 6),
    }
}

fn gen_fuzz_enum(rng: &mut Rng) -> FuzzEnum {
    match rng.range(4) {
        0 => FuzzEnum::Unit,
        1 => FuzzEnum::U64(rng.next_u64()),
        2 => FuzzEnum::Bytes(gen_arr4(rng)),
        _ => FuzzEnum::Pair {
            a: rng.next_u16(),
            b: rng.next_u8(),
        },
    }
}

fn gen_entry_map_u8_u16(rng: &mut Rng, max_len: usize) -> EntryMapU8U16 {
    let len = rng.range(max_len + 1);
    EntryMapU8U16 {
        entries: (0..len)
            .map(|_| ((rng.next_u8() % 5), rng.next_u16()))
            .collect(),
    }
}

fn gen_entry_map_string_u32(rng: &mut Rng, max_len: usize) -> EntryMapStringU32 {
    let len = rng.range(max_len + 1);
    EntryMapStringU32 {
        entries: (0..len)
            .map(|_| (gen_ascii_string(rng, 4), rng.next_u32()))
            .collect(),
    }
}

fn main() {
    let mut ser_rng = Rng::new(SEED);

    for i in 0..SER_CASES {
        emit_bytes(
            &format!("ser_bool_{i:03}"),
            &bcs::to_bytes(&ser_rng.next_bool()).unwrap(),
        );
        emit_bytes(
            &format!("ser_u64_{i:03}"),
            &bcs::to_bytes(&ser_rng.next_u64()).unwrap(),
        );
        emit_bytes(
            &format!("ser_i64_{i:03}"),
            &bcs::to_bytes(&(ser_rng.next_u64() as i64)).unwrap(),
        );
        emit_bytes(
            &format!("ser_string_{i:03}"),
            &bcs::to_bytes(&gen_ascii_string(&mut ser_rng, 12)).unwrap(),
        );
        emit_bytes(
            &format!("ser_bytes_{i:03}"),
            &bcs::to_bytes(&gen_bytes(&mut ser_rng, 16)).unwrap(),
        );
        emit_bytes(
            &format!("ser_vec_u16_{i:03}"),
            &bcs::to_bytes(&gen_u16_vec(&mut ser_rng, 8)).unwrap(),
        );
        emit_bytes(
            &format!("ser_vec_vec_u8_{i:03}"),
            &bcs::to_bytes(&gen_vec_vec_u8(&mut ser_rng, 4, 8)).unwrap(),
        );
        emit_bytes(
            &format!("ser_opt_u64_{i:03}"),
            &bcs::to_bytes(&gen_opt_u64(&mut ser_rng)).unwrap(),
        );
        emit_bytes(
            &format!("ser_struct_{i:03}"),
            &bcs::to_bytes(&gen_fuzz_struct(&mut ser_rng)).unwrap(),
        );
        emit_bytes(
            &format!("ser_enum_{i:03}"),
            &bcs::to_bytes(&gen_fuzz_enum(&mut ser_rng)).unwrap(),
        );
        emit_bytes(
            &format!("ser_map_u8_u16_{i:03}"),
            &bcs::to_bytes(&gen_entry_map_u8_u16(&mut ser_rng, 6)).unwrap(),
        );
        emit_bytes(
            &format!("ser_map_string_u32_{i:03}"),
            &bcs::to_bytes(&gen_entry_map_string_u32(&mut ser_rng, 6)).unwrap(),
        );
        emit_bytes(
            &format!("ser_deep_struct_{i:03}"),
            &bcs::to_bytes(&gen_deep_struct(&mut ser_rng)).unwrap(),
        );
    }

    let mut de_rng = Rng::new(SEED ^ 0x9e37_79b9_7f4a_7c15);

    for i in 0..DE_CASES {
        let bytes = gen_bytes(&mut de_rng, 24);
        emit_status::<bool>(&format!("de_bool_{i:03}"), &bytes);
        emit_status::<Option<u8>>(&format!("de_opt_u8_{i:03}"), &bytes);
        emit_status::<String>(&format!("de_string_{i:03}"), &bytes);
        emit_status::<Vec<u8>>(&format!("de_vec_u8_{i:03}"), &bytes);
        emit_status::<Vec<u16>>(&format!("de_vec_u16_{i:03}"), &bytes);
        emit_status::<FuzzEnum>(&format!("de_enum_{i:03}"), &bytes);
        emit_status::<(u32, bool)>(&format!("de_tuple_pair_{i:03}"), &bytes);
        emit_status::<BTreeMap<u8, ()>>(&format!("de_map_u8_unit_{i:03}"), &bytes);
        emit_status::<FuzzStruct>(&format!("de_struct_{i:03}"), &bytes);
        emit_status::<DeepStruct>(&format!("de_deep_struct_{i:03}"), &bytes);
        emit_status::<BTreeMap<u8, u16>>(&format!("de_map_u8_u16_{i:03}"), &bytes);
        emit_status::<BTreeMap<String, u32>>(&format!("de_map_string_u32_{i:03}"), &bytes);
    }
}
